### ISSUE TITLE
duplicate column names are mutated in place

### ISSUE THREAD

**Issue by glentakahashi** [duckdb/duckdb#2365]
**What does happen?**
In DuckDB 0.2.9, (https://github.com/duckdb/duckdb/pull/2214) registered dataframes that contain duplicate column names are automatically renamed to be `a, a_1, a_2` etc. This rename isn't localized to DuckDB, and pandas dataframes will be modified in place.

**What should happen?**
DuckDB should not modify the dataframe on registration unless an option to automatically rename columns is specified. 

This is dangerous because before, if a user tried to register a column with duplicates, it would crash and force the user to decide how they want to handle duplicates so that they can update the rest of the code that references the dataframe as well. Now, if you accidentally register a dataframe with duplicate columns, no indication or warning is displayed, and other parts of the code that depend on those columns may now break.

**Comment by Mytherin** [comment]
The problem here is not that duplicate column names are handled, but rather that `column_attributes` is modified, which modifies the Pandas DataFrame that is passed in. I don't think we need to have a flag there or throw an exception, we just need to not modify the `column_attributes` of the Pandas DataFrame.

**Comment by pdet** [comment]
Ah, I misread it, sure, I agree there is no reason to modify the original DF.

**Comment by jlorince** [comment]
I thought this was fixed, but just confirmed that the problem persists in 0.3.1. Minimal repro:
```
In [11]: import duckdb
    ...: print(duckdb.__version__)
    ...: dupecols = pandas.DataFrame({"a":[1,2,3], "b":[1,2,3]})
    ...: dupecols.columns = ["a", "a"]
    ...: con = duckdb.connect(database=":memory:", read_only=False)
    ...: con.register("dupecols", dupecols)
    ...: print(dupecols)
    ...:
0.3.1
   a  a_1
0  1    1
1  2    2
2  3    3
```

Any ideas what happened? CC @glentakahashi @Mytherin

**Comment by hannesmuehleisen** [comment]
@pdet ?

**Comment by TNonet** [comment]
@pdet This still seems to be happening in `0.3.4`

Using @jlorince's code from above.
```
import duckdb
print(duckdb.__version__)
dupecols = pandas.DataFrame({"a":[1,2,3], "b":[1,2,3]})
dupecols.columns = ["a", "a"]
con = duckdb.connect(database=":memory:", read_only=False)
con.register("dupecols", dupecols)
print(dupecols)
0.3.4
   a  a_1
0  1    1
1  2    2
2  3    3
```

I believe this is due to reference vs. assignment of numpy values. For example, modifying the columns in place functionally "renames" both dataframes as both columns attribute references the same numpy array.
```
df =  pandas.DataFrame.from_dict({'a': [42], 'b': [32]})
df_copy = df.copy(True)
df.columns.values[0] = 'c'

print((df.columns == ['c', 'b']).all())
print((df_copy.columns == ['c', 'b']).all())
True
True
``` 

Where assigning a new list as the columns of the original df does not influence the copy.
```
df =  pandas.DataFrame.from_dict({'a': [42], 'b': [32]})
df_copy = df.copy(True)
df.columns = ['c', 'b']

print((df.columns == ['c', 'b']).all())
print((df_copy.columns == ['a', 'b']).all())
True
True
```

This test implemented in [#2403](https://github.com/duckdb/duckdb/pull/2403/files) should check explicitly for values.

```
import pandas as pd
import duckdb

def test_multiple_columns_with_same_name():
    df = pd.DataFrame({'a': [1, 2, 3, 4], 'b': [5, 6, 7, 8], 'd': [9, 10, 11, 12]})
    df = df.rename(columns={ df.columns[1]: "a" })
    df_original = df.copy(True)
    con = duckdb.connect()
    con.register('df_view', df)

    assert con.execute("DESCRIBE df_view;").fetchall() == [('a', 'BIGINT', 'YES', None, None, None), ('a_1', 'BIGINT', 'YES', None, None, None), ('d', 'BIGINT', 'YES', None, None, None)]
    assert con.execute("select a_1 from df_view;").fetchall() == [(5,), (6,), (7,), (8,)]
    assert con.execute("select a from df_view;").fetchall() == [(1,), (2,), (3,), (4,)]
    # Verify we are not changing original dataframe
    assert all(df_original.columns == ['a', 'a', 'd']), df_original.columns

test_multiple_columns_with_same_name()
0.3.4
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-28-3137430a102e> in <cell line: 18>()
     16 
     17 print(duckdb.__version__)
---> 18 test_multiple_columns_with_same_name()

<ipython-input-28-3137430a102e> in test_multiple_columns_with_same_name()
     13     assert con.execute("select a from df_view;").fetchall() == [(1,), (2,), (3,), (4,)]
     14     # Verify we are not changing original dataframe
---> 15     assert all(df_original.columns == ['a', 'a', 'd']), df_original.columns
     16 
     17 print(duckdb.__version__)

AssertionError: Index(['a', 'a_1', 'd'], dtype='object')
```
